--- a/include/configs/stm32mp13_common.h	2024-09-30 18:38:41.388049737 +0900
+++ b/include/configs/stm32mp13_common.h	2024-09-30 18:39:56.180092954 +0900
@@ -100,7 +100,6 @@
 	"if test ${boot_device} = serial || test ${boot_device} = usb;" \
 	"then stm32prog ${boot_device} ${boot_instance}; " \
 	"else " \
-		"run env_check;" \
 		"if test ${boot_device} = mmc;" \
 		"then env set boot_targets \"mmc${boot_instance}\"; fi;" \
 		"if test ${boot_device} = nand ||" \
@@ -110,6 +109,7 @@
 	"fi;\0"
 
 #define STM32MP_EXTRA \
+	"run env_check;" \
 	"env_check=if env info -p -d -q; then env save; fi\0" \
 	"boot_net_usb_start=true\0"
 
