--- a/core/arch/arm/mm/core_mmu.c  2024-09-16 20:14:20.510616345 +0900
+++ b/core/arch/arm/mm/core_mmu.c  2024-09-16 20:14:51.001753746 +0900
@@ -1399,7 +1399,7 @@
 /*
  * teecore gets some memory area definitions
  */
-void core_mmu_get_mem_by_type(unsigned int type, vaddr_t *s, vaddr_t *e)
+void core_mmu_get_mem_by_type(enum teecore_memtypes type, vaddr_t *s, vaddr_t *e)
 {
        struct tee_mmap_region *map = find_map_by_type(type);

