--- a/ta/pkcs11/src/processing.h	2024-09-22 16:20:38.134594408 +0900
+++ b/ta/pkcs11/src/processing.h	2024-09-22 16:21:09.235661744 +0900
@@ -107,7 +107,7 @@
 /*
  * Symmetric crypto algorithm specific functions
  */
-bool processing_is_tee_symm(uint32_t proc_id);
+bool processing_is_tee_symm(enum pkcs11_mechanism_id proc_id);
 
 enum pkcs11_rc init_symm_operation(struct pkcs11_session *session,
 				   enum processing_func function,
